In [1]:
import csv
import os
import math

In [2]:
def media(x_lst: list[float]) -> float :
    return sum(x_lst) / len(x_lst)

def pob_desv_std(x_lst: list[float]) -> float:
    mu = media(x_lst)
    return math.sqrt(sum([(x - mu)**2 for x in x_lst]) / len(x_lst))

In [3]:
file_path = os.path.join('..','Datasets','iris','iris.csv')

with open(file_path) as f:
    sepal_length_setosa, sepal_width_setosa, petal_length_setosa, petal_width_setosa = [], [], [], []
    sepal_length_versicolor, sepal_width_versicolor, petal_length_versicolor, petal_width_versicolor = [], [], [], []
    sepal_length_virginica, sepal_width_virginica, petal_length_virginica, petal_width_virginica = [], [], [], []
    for row in csv.reader(f):
        if len(row) != 0:
            if row[4] == 'Iris-setosa':
                sepal_length_setosa.append(float(row[0]))
                sepal_width_setosa.append(float(row[1]))
                petal_length_setosa.append(float(row[2]))
                petal_width_setosa.append(float(row[3]))
            elif row[4] == 'Iris-versicolor':
                sepal_length_versicolor.append(float(row[0]))
                sepal_width_versicolor.append(float(row[1]))
                petal_length_versicolor.append(float(row[2]))
                petal_width_versicolor.append(float(row[3]))
            elif row[4] == 'Iris-virginica':
                sepal_length_virginica.append(float(row[0]))
                sepal_width_virginica.append(float(row[1]))
                petal_length_virginica.append(float(row[2]))
                petal_width_virginica.append(float(row[3]))

In [4]:
sepal_length = sepal_length_setosa + sepal_length_versicolor + sepal_length_virginica
sepal_width = sepal_width_setosa + sepal_width_versicolor + sepal_width_virginica
petal_length = petal_length_setosa + petal_length_versicolor + petal_length_virginica
petal_width = petal_width_setosa + petal_width_versicolor + petal_width_virginica

In [5]:
mean_sepal_length = media(sepal_length)
mean_sepal_width = media(sepal_width)
mean_petal_length = media(petal_length)
mean_petal_width = media(petal_width)

pstdev_sepal_length = pob_desv_std(sepal_length)
pstdev_sepal_width = pob_desv_std(sepal_width)
pstdev_petal_length = pob_desv_std(petal_length)
pstdev_petal_width = pob_desv_std(petal_width)

mean_sepal_length_setosa = media(sepal_length_setosa)
mean_sepal_width_setosa = media(sepal_width_setosa)
mean_petal_length_setosa = media(petal_length_setosa)
mean_petal_width_setosa = media(petal_width_setosa)

pstdev_sepal_length_setosa = pob_desv_std(sepal_length_setosa)
pstdev_sepal_width_setosa = pob_desv_std(sepal_width_setosa)
pstdev_petal_length_setosa = pob_desv_std(petal_length_setosa)
pstdev_petal_width_setosa = pob_desv_std(petal_width_setosa)

mean_sepal_length_versicolor = media(sepal_length_versicolor)
mean_sepal_width_versicolor = media(sepal_width_versicolor)
mean_petal_length_versicolor = media(petal_length_versicolor)
mean_petal_width_versicolor = media(petal_width_versicolor)

pstdev_sepal_length_versicolor = pob_desv_std(sepal_length_versicolor)
pstdev_sepal_width_versicolor = pob_desv_std(sepal_width_versicolor)
pstdev_petal_length_versicolor = pob_desv_std(petal_length_versicolor)
pstdev_petal_width_versicolor = pob_desv_std(petal_width_versicolor)

mean_sepal_length_virginica = media(sepal_length_virginica)
mean_sepal_width_virginica = media(sepal_width_virginica)
mean_petal_length_virginica = media(petal_length_virginica)
mean_petal_width_virginica = media(petal_width_virginica)

pstdev_sepal_length_virginica = pob_desv_std(sepal_length_virginica)
pstdev_sepal_width_virginica = pob_desv_std(sepal_width_virginica)
pstdev_petal_length_virginica = pob_desv_std(petal_length_virginica)
pstdev_petal_width_virginica = pob_desv_std(petal_width_virginica)

In [6]:
from IPython.display import display, Markdown

markdown_text = f"""
# Resumen de medidas de las flores Iris (cm)

## Medidas generales

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | {mean_sepal_length:.1f}   | {pstdev_sepal_length:.1f}                 |
| Ancho sépalo           | {mean_sepal_width:.1f}    | {pstdev_sepal_width:.1f}                  |
| Longitud pétalo        | {mean_petal_length:.1f}   | {pstdev_petal_length:.1f}                 |
| Ancho pétalo           | {mean_petal_width:.1f}    | {pstdev_petal_width:.1f}                  |

## Medidas por especie

### Iris-setosa

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | {mean_sepal_length_setosa:.1f}   | {pstdev_sepal_length_setosa:.1f}                 |
| Ancho sépalo           | {mean_sepal_width_setosa:.1f}    | {pstdev_sepal_width_setosa:.1f}                  |
| Longitud pétalo        | {mean_petal_length_setosa:.1f}   | {pstdev_petal_length_setosa:.1f}                 |
| Ancho pétalo           | {mean_petal_width_setosa:.1f}    | {pstdev_petal_width_setosa:.1f}                  |

### Iris-versicolor

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | {mean_sepal_length_versicolor:.1f}   | {pstdev_sepal_length_versicolor:.1f}                 |
| Ancho sépalo           | {mean_sepal_width_versicolor:.1f}    | {pstdev_sepal_width_versicolor:.1f}                  |
| Longitud pétalo        | {mean_petal_length_versicolor:.1f}   | {pstdev_petal_length_versicolor:.1f}                 |
| Ancho pétalo           | {mean_petal_width_versicolor:.1f}    | {pstdev_petal_width_versicolor:.1f}                  |

### Iris-virginica

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | {mean_sepal_length_virginica:.1f}   | {pstdev_sepal_length_virginica:.1f}                 |
| Ancho sépalo           | {mean_sepal_width_virginica:.1f}    | {pstdev_sepal_width_virginica:.1f}                  |
| Longitud pétalo        | {mean_petal_length_virginica:.1f}   | {pstdev_petal_length_virginica:.1f}                 |
| Ancho pétalo           | {mean_petal_width_virginica:.1f}    | {pstdev_petal_width_virginica:.1f}                  |

Nota: Las medidas presentadas están a una cifra después del punto decimal ya que los datos originales están reportados con dicha cantidad de cifras.
"""

display(Markdown(markdown_text))


# Resumen de medidas de las flores Iris (cm)

## Medidas generales

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | 5.8   | 0.8                 |
| Ancho sépalo           | 3.1    | 0.4                  |
| Longitud pétalo        | 3.8   | 1.8                 |
| Ancho pétalo           | 1.2    | 0.8                  |

## Medidas por especie

### Iris-setosa

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | 5.0   | 0.3                 |
| Ancho sépalo           | 3.4    | 0.4                  |
| Longitud pétalo        | 1.5   | 0.2                 |
| Ancho pétalo           | 0.2    | 0.1                  |

### Iris-versicolor

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | 5.9   | 0.5                 |
| Ancho sépalo           | 2.8    | 0.3                  |
| Longitud pétalo        | 4.3   | 0.5                 |
| Ancho pétalo           | 1.3    | 0.2                  |

### Iris-virginica

| Característica          | Media | Desviación estándar |
|-------------------------|-------|---------------------|
| Longitud sépalo        | 6.6   | 0.6                 |
| Ancho sépalo           | 3.0    | 0.3                  |
| Longitud pétalo        | 5.6   | 0.5                 |
| Ancho pétalo           | 2.0    | 0.3                  |

Nota: Las medidas presentadas están a una cifra después del punto decimal ya que los datos originales están reportados con dicha cantidad de cifras.


#### PREGUNTA: 
¿Cuál de las 4 mediciones considerarías como la más característica para las tres especies? En otras palabras, ¿qué medición considerarías “mejor”, si tuvieras que adivinar la especie de Iris basándote solo en esos cuatro valores?

#### RESPUESTA:
Basándonos en las medidas generales nos interesa aquella que tenga mayor dispersión. Esto nos indica qué tan separadas están las medidas entre cada una de las especies, o dicho de otra forma, hay menos superposición entre mediciones. Así mismo, considerando las medidas para cada una de las especies queremos aquella que tenga menor dispersión, pues queremos adivinar la especie con un pequeño margen de error.

Observamos que la medida que cumple con lo descrito es el **ancho del pétalo** ya que la desviación estándar (en cm) es:

* **En general:** 0.8
* **Iris-setosa:** 0.1
* **Iris-versicolor:** 0.2
* **Iris-virginica:** 0.3

Si consideramos la media y la disviación estándar para crear un intervalo aproximado de mediciones (en cm), tenemos lo siguiente:

* **Iris-setosa:** (0.2 ± 0.1) → (0.1, 0.3)
* **Iris-versicolor:** (1.3 ± 0.2) → (1.1, 1.5)
* **Iris-virginica:** (2.0 ± 0.3) → (1.7, 2.3)

Observamos que no hay superposición entre las medidas del **ancho del pétalo**, pues es lo que buscábamos al elegir aquella medida general con gran dispersión. Por otro lado, entre *iris-versicolor* e *iris-virginica* hay una separación de apenas 0.2 cm entre sus intervalos. Si hubiera valores extremos en los datos, algunos *iris-versicolor* grandes (\~1.5 cm) podrían confundirse con *iris-virginica* pequeños (\~1.7 cm). Es por ello que elegimos aquella medida que entre especies tenga menor dispersión: para que el margen de error sea pequeño y que los datos estén más cercanos a la media.